src/conversions/array.rs

```diff
@@ -0,0 +1,72 @@
+use crate::{exceptions, types::PySequence, FromPyObject, PyAny, PyResult, PyTryFrom};
+
+// Helper to safely create arrays since the standard library doesn't
+// provide one yet. Shouldn't be necessary in the future.
+#[cfg(min_const_generics)]
+struct ArrayGuard<T, const N: usize> {
+    dst: *mut T,
+    initialized: usize,
+}
+
+#[cfg(min_const_generics)]
+impl<T, const N: usize> Drop for ArrayGuard<T, N> {
+    fn drop(&mut self) {
+        debug_assert!(self.initialized <= N);
+        let initialized_part = core::ptr::slice_from_raw_parts_mut(self.dst, self.initialized);
+        unsafe {
+            core::ptr::drop_in_place(initialized_part);
+        }
+    }
+}
+
+#[cfg(min_const_generics)]
+pub fn create_array_from_obj<'s, T, const N: usize>(obj: &'s PyAny) -> PyResult<[T; N]>
+where
+    T: FromPyObject<'s>,
+{
+    let seq = <PySequence as PyTryFrom>::try_from(obj)?;
+    try_create_array(|idx| {
+        seq.get_item(idx as isize)
+            .map_err(|_| {
+                exceptions::PyBufferError::new_err("Slice length does not match buffer length.")
+            })?
+            .extract::<T>()
+    })
+}
+
+#[cfg(any(all(min_const_generics, feature = "nightly"), not(min_const_generics)))]
+pub fn extract_sequence_into_slice<'s, T>(obj: &'s PyAny, slice: &mut [T]) -> PyResult<()>
+where
+    T: FromPyObject<'s>,
+{
+    let seq = <PySequence as PyTryFrom>::try_from(obj)?;
+    if seq.len()? as usize != slice.len() {
+        return Err(exceptions::PyBufferError::new_err(
+            "Slice length does not match buffer length.",
+        ));
```

Comment:
As this particular error is now duplicated in a lot of places, I would like to see this refactored into a standalone function - e.g. `invalid_sequence_length()`.

For bonus points, it would be great to improve the error at the same time:
- it's not involving the Python buffer protocol when casting sequences to arrays in this way, so `ValueError` & `sequence` seems more appropriate terminology
- add size information which would help the user

e.g. maybe this:

```
fn invalid_sequence_length(expected: usize, actual: usize) -> PyErr {
    exceptions::PyValueError::new_err(
        format!("expected a sequence of length {} (got {})", expected, actual)
    )
}
```

Happy for you to go with something different if you think alternative wording is clearer.